source/h2_stream.c

```diff
@@ -535,9 +536,10 @@ int aws_h2_stream_on_decoder_push_promise(struct aws_h2_stream *stream, uint32_t
         s_get_h2_connection(stream), promised_stream_id, AWS_H2_ERR_REFUSED_STREAM);
 }
 
-int aws_h2_stream_on_decoder_data(struct aws_h2_stream *stream, struct aws_byte_cursor data) {
+int aws_h2_stream_on_decoder_data_begin(struct aws_h2_stream *stream, uint32_t data_payload_len) {
     AWS_PRECONDITION_ON_CHANNEL_THREAD(stream);
 
+    stream->thread_data.window_size_self -= data_payload_len;
     if (s_check_state_allows_frame_type(stream, AWS_H2_FRAME_T_DATA)) {
```

Comment:
Don't alter the stream until we've checked whether it's even legal to receive the DATA frame

Also, if stream's window_size goes below 0 we need to RST_STREAM with FLOW_CONTROL_ERROR